src/hotspot/share/gc/parallel/parallelScavengeHeap.cpp

```diff
@@ -549,8 +549,28 @@ void ParallelScavengeHeap::collect(GCCause::Cause cause) {
     return;
   }
 
-  VM_ParallelGCSystemGC op(gc_count, full_gc_count, cause);
-  VMThread::execute(&op);
+  while (true) {
+    VM_ParallelGCSystemGC op(gc_count, full_gc_count, cause);
+    VMThread::execute(&op);
+
+    if (!GCCause::is_explicit_gc(cause) ||
+        !VM_ParallelGCSystemGC::is_cause_full(cause) ||
+        op.full_gc_succeeded()) {
+      return;
```

Comment:
Looking at the caller context, I'd believe all callers expect a GC cycle has just run (by this thread or another) when `collect(cause)` returns. Therefore, the terminating condition should be to check #gc_cycles against the intended gc-type (young or not), sth like:

```cpp
{
  MutexLocker ml(Heap_lock);
  if (is_young_gc) {
    if (gc_count != total_collections()) {
      return;
    }
  } else {
    if (full_gc_count != total_full_collections()) {
      return;
    }
  }
}
```

(Originally, I was thinking only about full-gc, i.e. systemgc and `_wb_full_gc`, but it also seems reasonable to provide such guarantee, a gc-cycle has run, for `_wb_young_gc`.)